### PR TITLE
Refactoring v3

### PR DESCRIPTION
**Pull request by ferringb**
Roughly:
* lock `dep` dependencies to what we care about and bump our deps to modern versions.
  * I'd suggest ignoring /vendor directory in review- all updates there are just from the dep's being upgraded.

  * This also upgrades our cfssl client lib to 1.3.3; that in turn means we drop go 1.9/1.10 support since cfssl no longer validates against those versions (and 1.9 has build failures in that lib).
  * Remove obsolete gopkg.in/yaml.v1 usage since we already have v.2, update dep accordingly.
* Make use of v.2 to enforce yaml strictness for config parsing.  If we don't recognize a value it's now an error.
* Split cert/cert.go into cert/{spec,ca,util}.go since it was getting a bit messy.
* Clean out remaining bad `range` usage- roughly, don't iterate the index of a range when you just want to iterate the items.
* Simplify metrics endpoint / further; drop the ability to supply arbitrary html blurb, drop static cert list.  For the extra html this isn't the place for that.  For the cert list, metrics is a far richer source for that- thus use that.
* Move CheckDiskPKI to Spec.  The goal for encapsulation here is thus:

  Spec: Manages PKI on disk
  CertSpecManager: manages triggering actions based on Spec events
  Manager: manages CertSpecManager instances, waking them as needed.

  Aligning with that allows us to remove the mtime checks strewn through out the code and centralize them (or replace them with x509 comparisons), thus removing more sources of errant restart/actions.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -37,10 +37,6 @@ has the following endpoints:
 * `/debug/pprof` is a
   [net/http/pprof](https://golang.org/pkg/net/http/pprof/) endpoint.
 
-In the configuration file, the `index_extra_html` may contain HTML
-that will be inserted at the top of the file, after the server start
-time and current host:port.
-
 ## Metrics
 
 The following metrics are collected by Prometheus:
```
Comment (anita-tenjarla) [inline README.md]: We should probably update this part of the readme since it's now outdated. And other stuff in the readme. I'm happy to do that in a separate pr if you'd like.

File: cert/ca.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := remote.Info(serialisedRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	return []byte(strings.TrimSpace(resp.Certificate)), nil
```
Comment (anita-tenjarla) [inline cert/ca.go]: I'm 99% sure the trim space here isn't necessary anymore. That's on me though -- I should've changed this earlier.

File: README.md (hunk 1 above)
Comment (ferringb) [reply README.md]: I'm kind of inclined to point the metrics doc section at the godoc page for the metrics package; maintaining the metrics text in two spots is annoying.

I'd say the same for the configs, but that needs to be refactored to support that.

Either way, if you don't mind- a followup PR would be appreciated.  Next change in the PR pipeline from me is lint/vet fixes, and then back to the actual bugremoval work.

File: cert/ca.go (hunk 2 above)
Comment (ferringb) [reply cert/ca.go]: My intent actually is to just leverage the hell out of that x509 comparison logic you added, and drop all of this- instead just defer to doing comparisons via that.  I was intending on doing the same thing for the mtime checks- we don't care if the spec is newer then cert/key on disk as long as x509 and key material is the same.  Which removes yet another errant restart....

Until then, I'd prefer we leave the trim alone- the code's going to die, and that pathway is a known source of incorrect restarts.

File: README.md (hunk 1 above)
Comment (anita-tenjarla) [reply README.md]: Sure thing, I'll make a readme PR Monday morning.
